ksql-engine/src/main/java/io/confluent/ksql/physical/PhysicalPlanBuilder.java

```diff
@@ -179,14 +205,38 @@ private QueryMetadata buildPlanForStructuredOutputNode(String sqlExpression, fin
               schemaKStream.getKeyField(),
               outputNode.getTimestampField(),
               outputNode.getKsqlTopic());
-
     }
 
-    if (updateMetastore) {
+    if (updateMetastore && outputNode.isDoCreateInto()) {
       metaStore.putSource(sinkDataSource.cloneWithTimeKeyColumns());
+    } else {
+      StructuredDataSource structuredDataSource = metaStore.getSource(sinkDataSource.getName());
+      if (structuredDataSource.getDataSourceType() != sinkDataSource.getDataSourceType()) {
+        throw new KsqlException(String.format("Incompatible data sink and query result. Data sink"
+                                              + " (%s)"
+                                              + " type is %s but select query result is %s.",
+                                              sinkDataSource.getName(), sinkDataSource
+                                                  .getDataSourceType(), structuredDataSource.getDataSourceType()));
+      }
+      Schema resultSchema = SchemaUtil.removeImplicitRowTimeRowKeyFromSchema(sinkDataSource
+                                                                                 .cloneWithTimeKeyColumns().getSchema());
+      if (!SchemaUtil.areEqualSchemas(resultSchema,
+                                      SchemaUtil.removeImplicitRowTimeRowKeyFromSchema(structuredDataSource.getSchema()))) {
+        throw new KsqlException(String.format("Incompatible schema between results and sink. "
+                                              + "Result schema is %s, but the sink schema is %s"
+                                              + ".", SchemaUtil.schemaString(resultSchema), SchemaUtil
+            .schemaString(structuredDataSource.getSchema())));
+      }
+      enforceKeyEquivalence(structuredDataSource.getKeyField(), sinkDataSource.getKeyField());
+
     }
 
-    final QueryId queryId = sinkDataSource.getPersistentQueryId();
+    final QueryId queryId;
+    if (outputNode.isDoCreateInto()) {
+      queryId = sinkDataSource.getPersistentQueryId();
+    } else {
+      queryId = new QueryId("InsertQuery_" + queryIdGenerator.getNextId());
```

Comment:
do you think it makes sense to include the stream name here, to keep it kind of consistent with CSAS and make for a more informative name?